clients/filesystem-fuse/src/memory_filesystem.rs

```diff
@@ -0,0 +1,250 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+use crate::filesystem::{FileReader, FileStat, FileWriter, PathFileSystem, Result};
+use crate::opened_file::{OpenFileFlags, OpenedFile};
+use async_trait::async_trait;
+use bytes::Bytes;
+use fuse3::FileType::{Directory, RegularFile};
+use fuse3::{Errno, FileType};
+use std::collections::BTreeMap;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex, RwLock};
+
+// Simple in-memory file implementation of MemoryFileSystem
+struct MemoryFile {
+    kind: FileType,
+    data: Arc<Mutex<Vec<u8>>>,
+}
+
+// MemoryFileSystem is a simple in-memory filesystem implementation
+// It is used for testing purposes
+pub struct MemoryFileSystem {
+    // file_map is a map of file name to file size
+    file_map: RwLock<BTreeMap<PathBuf, MemoryFile>>,
+}
+
+impl MemoryFileSystem {
+    const FS_META_FILE_NAME: &'static str = "/.gvfs_meta";
+
+    pub(crate) async fn new() -> Self {
+        Self {
+            file_map: RwLock::new(Default::default()),
+        }
+    }
+
+    fn create_file_stat(&self, path: &Path, file: &MemoryFile) -> FileStat {
+        match file.kind {
+            Directory => FileStat::new_dir_filestat_with_path(path),
+            _ => {
+                FileStat::new_file_filestat_with_path(path, file.data.lock().unwrap().len() as u64)
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl PathFileSystem for MemoryFileSystem {
+    async fn init(&self) -> Result<()> {
+        let root_file = MemoryFile {
+            kind: Directory,
+            data: Arc::new(Mutex::new(Vec::new())),
+        };
+        let root_path = PathBuf::from("/");
+        self.file_map.write().unwrap().insert(root_path, root_file);
+
+        let meta_file = MemoryFile {
+            kind: RegularFile,
+            data: Arc::new(Mutex::new(Vec::new())),
+        };
+        let meta_file_path = Path::new(Self::FS_META_FILE_NAME).to_path_buf();
+        self.file_map
+            .write()
+            .unwrap()
+            .insert(meta_file_path, meta_file);
+        Ok(())
+    }
+
+    async fn stat(&self, path: &Path) -> Result<FileStat> {
+        self.file_map
+            .read()
+            .unwrap()
+            .get(path)
+            .map(|x| self.create_file_stat(path, x))
+            .ok_or(Errno::from(libc::ENOENT))
+    }
+
+    async fn read_dir(&self, path: &Path) -> Result<Vec<FileStat>> {
+        let file_map = self.file_map.read().unwrap();
+
+        let results: Vec<FileStat> = file_map
+            .iter()
+            .filter(|x| path_in_dir(path, x.0))
```

Comment:
please filter out the files in sub directory

Reply:
fix